drivers: i2c: i2c_stm32_v2: stability improvements

erian747 opened the pull request:
Due to various reported stability issues with the STM32 I2C V2 driver in interrupt mode, the controller part is rewritten in an attempt to closely follow documentation found in reference manual and errata sheets.

See https://github.com/zephyrproject-rtos/zephyr/issues/70077

A side effect should also be increased performance

This PR replaces the old draft at https://github.com/zephyrproject-rtos/zephyr/pull/82940

erian747 committed:
drivers: i2c_stm32_v2: Rework interrupt handler

- Rework transaction flow  to better match reference manual
and errata sheet for STM32 I2Cv2 peripheral
- Handle message transactions larger than 255 bytes in isr
- Combine stm32_i2c_msg_write and stm32_i2c_msg_read into
one common function stm32_i2c_irq_xfer when using interrupts

Signed-off-by: Erik Andersson <erian747@gmail.com>

erian747 committed:
drivers: i2c_stm32_v2: Remove obsolete functions

stm32_i2c_msg_read and stm32_i2c_msg_write
for interrupt mode is replaced by a common
function: stm32_i2c_irq_xfer

Signed-off-by: Erik Andersson <erian747@gmail.com>

erian747 committed:
drivers: i2c_stm32_v2: DMA fixup

Adapt DMA implementation to match changes
introduced by new interrupt handler

Signed-off-by: Erik Andersson <erian747@gmail.com>

erian747 committed:
drivers: i2c_stm32_v2: Bus error errata fix

Don't abort transaction on bus error in master mode
according to errata sheet

Signed-off-by: Erik Andersson <erian747@gmail.com>

erian747 committed:
drivers: i2c_stm32_v2: Refactor end of master transfer

The stm32_i2c_master_mode_end function can be
replaced with a simple irq disable and giving to
the device sync semaphore

Signed-off-by: Erik Andersson <erian747@gmail.com>

erwango reviewed (commented):
First of all, thanks @erian747 for the work on this driver and the time you're taking to share your improvements.

Before starting the review, a quick note from my part:
while I do believe this definitely brings important improvements to the current driver, we should also make sure this doesn't bring regression to others (not a surprise to you I guess).
Then, despite the issues you faced, this driver is used w/o trouble in number of applications, which tells that driver behavior depends on the use cases and also that it's difficult to reproduce all conditions in test environment.
So, on the hypothesis that everything can't be verified by testing, we should take special care in the review.

In order to make this PR progress at a reasonable speed, here are few things that will help us I the review
Is there a possibility to split this PR into smaller chunks that would be easier to swallow in reviews ?
If not, further splitting the PR in more single purpose commits would greatly help in the review process. You made efforts on that already but I had a quick look and it's still not easy to track changes. I know by experience that this is a tedious task, but this is really efficient (this usually also helps to find bugs).
Finally, would you be able to provide more indications on the tests that were made on this change, like the targets used, kind of tests (Zephyr tests/drivers ones for instance), ... ?

Last preliminary note on the change:
Drop of LL usage, even though they are justified, imply potential issue in further maintenance, as we lose the thin hw portability it brings. 
This concern depends on IPs, some are more or less stable across series. Fortunately, I2C seems relatively stable in that regard, but yet...
So when possible, let's try to minimize removal of LL.

erian747 commented:
Thanks for the comment
Regarding "regression to other" i completely agree, especially the "target" mode probably needs some testing and adjustments
DMA and non interrupt mode should ofcourse also be tested (even if i "believe" i haven't made any regression there)

What i have tested is master mode with irq on real hardware: STM32F071 and STM32H747 platforms with various types of sensors ranging from small messages with single byte data to messages with up to 1536 bytes and it seems solid
Would be glad if someone using target mode could run a few test after i have made a few changes i already now know about should be done

Regarding splitting commits in to smaller chunks, i understand what you mean but the complicated commit "Rework interrupt handler" which required deep insight into how the I2C IP works im not sure it helps splitting it, because the changes are tightly coupled together. What i know would really be helpful during a review is a specification in form of a state diagram, activity diagram etc describing how things suppose to work, i will look into that...

For LL usage, i saw there is a function LL_I2C_HandleTransfer() that can help removing a few direct register writes without having any impact on code, and also some writes to CR1 can be replaced by LL functions

I will update the PR as soon as i have time

erwango commented:
Heads up: You'll probably be (badly) hit by https://github.com/zephyrproject-rtos/zephyr/pull/87731...

erian747 commented:
Thanks for mentioning it, seems like its mostly renaming done in i2c_ll_stm32_v2.c so shouldn't be hard to manage

etienne-lms reviewed (commented):
The changes look sane to me.
Some minor comments.

Since there are now quite common few common functions between the irq and non-irq implementation, I wonder if it would help to have 3 source files. Well, maybe in a later change.

@erwango, with this change, there are a few raw manipulations of the I2C registers content (interrupt bits, CR1/CR2 fields loading, ...). I'm pretty fine with that but maybe you prefer we stick at most to the LL API? Since we're in a STM32 I2C v2 controller driver, all related SoCs have common fields/macros naming in the LL. It seems reliable to me (from a maintenance perspective).

erian747 commented:
> The changes look sane to me. Some minor comments.
> 
> Since there are now quite common few common functions between the irq and non-irq implementation, I wonder if it would help to have 3 source files. Well, maybe in a later change.
> 
> @erwango, with this change, there are a few raw manipulations of the I2C registers content (interrupt bits, CR1/CR2 fields loading, ...). I'm pretty fine with that but maybe you prefer we stick at most to the LL API? Since we're in a STM32 I2C v2 controller driver, all related SoCs have common fields/macros naming in the LL. It seems reliable to me (from a maintenance perspective).

I also think it would be a good idea to split up the file, right now it is a bit to big to be readable
But as you say, probably better to make a separate small PR for that

erian747 commented:
Very quick and ugly sketch
![image](https://github.com/user-attachments/assets/3c5f4d80-2b7e-4df0-ad22-ba0d38cb7541)
Good thing when drawing thigs like this is that i realized the condition marked with red ??? probably could/should never happen
But easy to get lost so im going to think a bit more about it before modifying the code

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -791,6 +825,185 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+		 *   has finalized its transfer and is waiting for restart
+		 * For both cases, a new start condition shall be sent
+		 */
+		__ASSERT_NO_MSG(((isr & I2C_ISR_TC) != 0U) || (msg->flags & I2C_MSG_RESTART));
```
```suggestion
		__ASSERT_NO_MSG(((isr & I2C_ISR_TC) != 0U) ||
				((msg->flags & I2C_MSG_RESTART) != 0U));

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -791,6 +825,185 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+		goto error;
+	}
+
+	if (msg->flags & I2C_MSG_STOP) {
```
```suggestion
	if ((msg->flags & I2C_MSG_STOP) != 0U) {

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -1248,6 +1461,10 @@ int i2c_stm32_transaction(const struct device *dev,
 			  struct i2c_msg msg, uint8_t *next_msg_flags,
 			  uint16_t periph)
 {
+	int ret = 0;
```
prefer with an empty line betwen local variable definitions and 1st instructions.

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -791,6 +825,185 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+	}
+#endif /* CONFIG_DCACHE && CONFIG_I2C_STM32_V2_DMA */
+	return 0;
+error:
```
Could you add 2 empty lines above:
```patch
 #endif /* CONFIG_DCACHE && CONFIG_I2C_STM32_V2_DMA */
+
 	return 0;
+
 error:

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -791,6 +825,185 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+	 */
+	if (k_sem_take(&data->device_sync_sem, K_MSEC(msg->len + 10)) != 0) {
+		is_timeout = true;
+	}
```
Call to `dma_stop()` is missing here. Despite it's brought back in a next commit through `dma_finish()`, the current commit should support DMA (to prevent later issue, e.g. when using `git bisect`).

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -150,92 +150,80 @@ static int configure_dma(struct stream const *dma, struct dma_config *dma_cfg,
 
-	if (LL_I2C_IsEnabledReloadMode(i2c)) {
-		LL_I2C_SetTransferSize(i2c, msg->len);
+	if (msg->flags & I2C_MSG_READ) {
```
```suggestion
	if ((msg->flags & I2C_MSG_READ) != 0U) {
```

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -150,92 +150,80 @@ static int configure_dma(struct stream const *dma, struct dma_config *dma_cfg,
-static inline void msg_init(const struct device *dev, struct i2c_msg *msg,
-			    uint8_t *next_msg_flags, uint16_t slave,
-			    uint32_t transfer)
+static inline void dma_xfer_start(const struct device *dev, struct i2c_msg *msg)
```
can remove the `inline` keyword. Not really useful.

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -150,92 +150,80 @@ static int configure_dma(struct stream const *dma, struct dma_config *dma_cfg,
+
+		if (configure_dma(&cfg->rx_dma, &data->dma_rx_cfg,
+					&data->dma_blk_cfg) != 0) {
+			LOG_ERR("Problem setting up RX DMA");
```
The error needs to be reported IMHO.
Maybe can be done in a later P-R, since this behavior was already there in original implementation.

etienne-lms reviewed (commented):
**~~More~~ Most** of the review comments below are related to coding style issues like 
s/`if (foo & mask)`/`if ((foo & mask) != 0U)`/.

Commit "*[drivers: i2c_stm32_v2: Remove obsolete functions](https://github.com/zephyrproject-rtos/zephyr/pull/89501/commits/3638ab6f51b570973d50a47458af9037eb77e95a)"* claims to only remove obsolete functions but it does more. I think it could only do that, the other change could be move in other related commits, already present in this commit series.

That said, those changes look good to me and address some corner cases as they claim to.

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -548,74 +548,109 @@ void i2c_stm32_event(const struct device *dev)
+			 * Data for a single transfer remains in buffer, set its length and
+			 * - If more messages follow and transfer direction for next message is
+			 *   same, keep reload on
+			 * - If direction change or current message is the last,
```
nitpicking:
```suggestion
			 * - If direction changes or current message is the last,

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -548,74 +548,109 @@ void i2c_stm32_event(const struct device *dev)
+				cr2 &= ~I2C_CR2_RELOAD;
 			}
-#endif /* CONFIG_I2C_STM32_V2_DMA */
+			/*
+			 * If reload was not disabled above, expect to arrive in this if/else
+			 * statement with data->current.len == 0
+			 */
```
I think you can remove this inline comment.

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -791,6 +826,200 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+	/* Clear fields in CR2 which will be filled in later in function */
+	cr2 &= ~(I2C_CR2_RELOAD | I2C_CR2_AUTOEND | I2C_CR2_NBYTES_Msk | I2C_CR2_SADD_Msk);
+
+	if (I2C_ADDR_10_BITS & data->dev_config) {
```
```suggestion
	if ((I2C_ADDR_10_BITS & data->dev_config) != 0U) {

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -791,6 +826,200 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+#endif
+
+#if defined(CONFIG_DCACHE) && defined(CONFIG_I2C_STM32_V2_DMA)
+	if (!buf_in_nocache((uintptr_t)msg->buf, msg->len) &&
```
Could you rather use `stm32_buf_in_nocache()` SoC helper function (here and below in this same function) in commit *"drivers: i2c_stm32_v2: Rework interrupt handler"*? You can remove implementation of local helper function `buf_in_nocache()` from commit *"drivers: i2c_stm32_v2: Remove obsolete functions"* that is anyway removed in the last commit of this series.

etienne-lms reviewed (commented):
Commit "drivers: i2c_stm32_v2: DMA fixup" fixes an regression introduced by a former commit in the series. It could be better all commits are consistent and add no regression (build and run) when possible. That helps a lot when digging for a regression using `git bisect` or like.
I would suggest to add insert a preparatory commit at the beginning of your patch series to introduce `dma_xfert_start()` and `dma_finish()` local helper functions and have `msg_init()` calling them.

erian747 committed:
drivers: i2c_stm32_v2: Add DMA helpers

To prepare for further patches and separation
of interrupt/DMA modes from polled, this commit
introduces two helpers for starting and stopping
DMA transfers

Finsishing DMA transfers are also moved from
interrupt handlers to thread after beeing wokenen
by semaphore

Signed-off-by: Erik Andersson <erian747@gmail.com>

erian747 committed:
drivers: i2c_stm32_v2: Abort on DMA error

If configuration of DMA is unsuccessful then
abort transaction and return with an error code

Signed-off-by: Erik Andersson <erian747@gmail.com>

erian747 commented:
Was a bit hard to move DMA changes to an initial commit, lots of conflicts with commits following
but i did give it a go, and atleast the resulting code after the final commit in the patch sequence matches with how it was before
so total behavior of this PR should be unaffected by latest push

erian747 commented:
New push again
Fix for DMA transfers when message size is > 255 bytes
I also extended the refactoring in commit "Refactor end of master transfer" as the stm32_i2c_irq_xfer function became quite long
So i moved the second part of the function into a new function "stm32_i2c_irq_msg_finish"

erwango commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -1295,7 +1536,6 @@ int i2c_stm32_transaction(const struct device *dev,
 	} while (rest > 0U);
 
-#ifndef CONFIG_I2C_STM32_INTERRUPT
 	struct i2c_stm32_data *data = dev->data;
 	const struct i2c_stm32_config *cfg = dev->config;
 	I2C_TypeDef *i2c = cfg->i2c;
```
Nitpick: those declarations should now go at the beginning of the block

etienne-lms commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -655,140 +630,205 @@ int i2c_stm32_error(const struct device *dev)
-			msg->buf, msg->len);
-		sys_cache_data_invd_range((void *)msg->buf, msg->len);
+#if defined(CONFIG_DCACHE) && defined(CONFIG_I2C_STM32_V2_DMA)
+	if (!stm32_buf_in_nocache((uintptr_t)msg->buf, msg->len) &&
+		((msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE)) {
```
```suggestion
	if (!stm32_buf_in_nocache((uintptr_t)msg->buf, msg->len) &&
	    ((msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE)) {

erian747 replied:
Fixed

etienne-lms reviewed (commented):
Some minor comments. I'm still looking at the overall changes.

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -150,6 +150,71 @@ static int configure_dma(struct stream const *dma, struct dma_config *dma_cfg,
+		if (data->current.len != 0U) {
+			/* Configure TX DMA */
+			data->dma_blk_cfg.source_address =
+				(uint32_t)data->current.buf;
```
Fixed

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -150,6 +150,71 @@ static int configure_dma(struct stream const *dma, struct dma_config *dma_cfg,
+		LL_I2C_DisableDMAReq_RX(cfg->i2c);
+#if defined(CONFIG_DCACHE)
+		if (!stm32_buf_in_nocache((uintptr_t)msg->buf, msg->len)) {
+			sys_cache_data_invd_range((void *)msg->buf, msg->len);
```
Fixed

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -566,63 +562,111 @@ void i2c_stm32_event(const struct device *dev)
-	}
+		data->current.len -= xfer_len;
+		data->current.buf += xfer_len;
+#endif
```
Fixed

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -548,74 +482,111 @@ void i2c_stm32_event(const struct device *dev)
+
+	} else if ((isr & I2C_ISR_RXNE) != 0U) {
+		__ASSERT_NO_MSG(data->current.len != 0U);
+		*data->current.buf = LL_I2C_ReadReg(regs, RXDR);
```
Fixed

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -548,74 +482,111 @@ void i2c_stm32_event(const struct device *dev)
+		/* Transfer Complete, (I2C_ISR_TC is set) no reload this time so either do
+		 * stop now or restart in thread
+		 */
+		__ASSERT_NO_MSG((isr & I2C_ISR_TC) != 0U);
```
Good point, fixed now

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -655,140 +630,205 @@ int i2c_stm32_error(const struct device *dev)
-		LOG_DBG("%s: NACK", __func__);
-		data->current.is_nack = 0U;
+	/*
+	 * For messages larger than 255 bytes, transactions must be split i chunks
```
Fixed

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -150,6 +150,71 @@ static int configure_dma(struct stream const *dma, struct dma_config *dma_cfg,
+	}
+}
+
+static inline void dma_finish(const struct device *dev, struct i2c_msg *msg)
```
Fixed

erian747 commented:
I found an issue today when using I2C write + DMA + more than 255 bytes, last byte was not sent
Made a new push and the added #ifdef at line 781 and enabling of DMA before starting transaction solves this issue 
Also fixed a few multi-line comments formatting

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -655,140 +626,202 @@ int i2c_stm32_error(const struct device *dev)
-		LOG_DBG("%s: ARLO %d", __func__,
-				    data->current.is_arlo);
-		data->current.is_arlo = 0U;
+	/* If this is not a stop message and more messages follow without change of direction,
```
Could very well be
Yes please test this branch and see if it solves your problem. If not then I gladly help out and fix it and it could be included in this PR as its in the same topic

cfriedt replied:
@erian747 - as far as I can tell, this works for us too 👍

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -794,6 +837,198 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
+	uint32_t isr = LL_I2C_ReadReg(regs, ISR);
+
+	/* Clear fields in CR2 which will be filled in later in function */
+	cr2 &= ~(I2C_CR2_RELOAD | I2C_CR2_AUTOEND | I2C_CR2_NBYTES_Msk | I2C_CR2_SADD_Msk);
```
Good catch, fixed

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -182,14 +183,14 @@ static void dma_xfer_start(const struct device *dev, struct i2c_msg *msg)
 			LL_I2C_EnableDMAReq_TX(i2c);
 		}
 	}
+	return ret;
```
Fixed

erian747 commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -563,63 +559,110 @@ void i2c_stm32_event(const struct device *dev)
+		/* Acknowledge stop condition */
+		LL_I2C_ClearFlag_STOP(regs);
+		/* Flush I2C controller TX buffer */
+		LL_I2C_ClearFlag_TXE(regs);
```
It should be taken care of in stm32_i2c_irq_xfer() when its called for next message to be transfered

cfriedt commented:
@danieldegrasse - you may want to review this change

cfriedt commented:
Darn - we experience some performance related issues that popped up later after using this change. I'll see if I can find out what the issue is.

cfriedt commented:
Hard to say what the issue is at this point. Timeout for i2c can definitely be a bit tricky as well.
